runtime/major_gc.c

```diff
@@ -910,7 +910,7 @@ Caml_noinline static intnat do_some_marking(struct mark_stack* stk,
     for (; me.start < scan_end; me.start++) {
       CAMLassert(budget >= 0);
 
-      value child = *me.start;
+      value child = atomic_load_relaxed_val(me.start);
```

Comment:
The discussion above is a bit hard to follow because concerns are mixed with TSan considerations, but my suggestion was to change:
```c
typedef struct {
  value* start;
  value* end;
} mark_entry; /* represents fields in the span [start, end) */
```
to:
```c
typedef struct {
  volatile value * start;
  volatile value * end;
} mark_entry; /* represents fields in the span [start, end) */
```
(At this point I'd introduce `typedef volatile value * value_ptr` to denote pointers into the OCaml heap in the runtime codebase.)

In this case we avoid introducing `atomic_load_relaxed_val`.

Would this fix be enough?

Reply:
I also think it's the way to go for correctness if we can't make them atomic.

Just to be sure, I think I've read in your #10992 PR or in one link provided as source that having volatile qualifier may have an impact on performance (preventing compiler optimisations).
This shouldn't be the case here, until we dereference the pointer, should it? `do_some_marking` plays a lot with `me.start` and we don't want to have a negative impact on performance.

Reply:
(You _could_ replace volatile with _Atomic and a relaxed load here, but use volatile for consistency with the rest of the codebase—it is better to discuss this style issue elsewhere such as #10992.)

It is the only read of this value, so there is nothing to optimize here. In addition, it is good to have in mind that the kind of optimization volatile/relaxed atomic prevents are in general not useful for code that is already carefully optimized by hand.

Note that making this change incites to add volatile to the definition of `Op_val`. This might be a thread to start unraveling, in relationship with #11426.